ratchet-tree.go

```diff
@@ -0,0 +1,479 @@
+	lhi := LeafNodeHashInput{HashType: 0}
+	if n.Node != nil {
+		p := n.Node.PublicKey
+		c := n.Node.Cred
+		if c == nil {
+			panic(fmt.Errorf("mls.rtn: Leaf node not provisioned with a credentialp"))
+		}
+		lhi.Info = &LeafNodeInfo{
+			PublicKey:  *p,
+			Credential: *c,
+		}
+	}
+
+	h, err := syntax.Marshal(lhi)
+	if err != nil {
+		panic(fmt.Errorf("mls.rtn: Marshal error %v", err))
+	}
+	n.Hash = cs.digest(h)
+
+}
+
+func (n *OptionalRatchetNode) merge(o *RatchetTreeNode) {
+	if n.Node == nil {
+		n.Node = o
+	} else {
+		n.Node.Merge(o)
+	}
+}
+
+func (n *OptionalRatchetNode) setHash(cs CipherSuite, l OptionalRatchetNode, r OptionalRatchetNode) {
+	phi := ParentNodeHashInput{HashType: 1}
+	if n.Node != nil {
+		phi.Info = &ParentNodeInfo{PublicKey: n.Node.PublicKey, UnmergedLeaves: n.Node.UnmergedLeaves}
+	}
+	phi.LeftHash = l.Hash
+	phi.RightHash = r.Hash
+	data, err := syntax.Marshal(phi)
+	if err != nil {
+		panic(fmt.Errorf("mls.rtn: set hash error %v", err))
+	}
+	n.Hash = cs.digest(data)
+}
+
+func (n *OptionalRatchetNode) hasPrivate() bool {
+	if n.Node != nil && n.Node.PrivateKey != nil {
+		return true
+	}
+	return false
+}
+
+//// RatchetTree Node
+type RatchetTreeNode struct {
+	Cred           *Credential `tls:"optional"`
+	PublicKey      *HPKEPublicKey
+	PrivateKey     *HPKEPrivateKey
+	UnmergedLeaves []leafIndex `tls:"head=2"`
+	CipherSuite    CipherSuite
+}
+
+func newRatchetTreeNode(cs CipherSuite, secret []byte) *RatchetTreeNode {
+	if len(secret) > 0 {
+		priv, err := cs.hpke().Derive(secret)
+		if err != nil {
+			panic(fmt.Errorf("hpke private key generation failed: %v", err))
+		}
+
+		return &RatchetTreeNode{
+			Cred:           nil,
+			PrivateKey:     &priv,
+			PublicKey:      &priv.PublicKey,
+			UnmergedLeaves: []leafIndex{},
+			CipherSuite:    cs,
+		}
+	}
+
+	return &RatchetTreeNode{
+		Cred:           nil,
+		PrivateKey:     nil,
+		PublicKey:      nil,
+		UnmergedLeaves: []leafIndex{},
+		CipherSuite:    cs,
+	}
+}
+
+func (n *RatchetTreeNode) Merge(o *RatchetTreeNode) {
+	if o.PublicKey != nil && !o.PublicKey.equals(n.PublicKey) {
+		n.PublicKey = o.PublicKey
+		n.PrivateKey = nil
+	}
+
+	if o.PrivateKey != nil {
+		n.PrivateKey = o.PrivateKey
+	}
+
+	// reset unmerged leaves since the node is being updated.
+	n.UnmergedLeaves = []leafIndex{}
+}
+
+func (n *RatchetTreeNode) AddUnmerged(l leafIndex) {
+	n.UnmergedLeaves = append(n.UnmergedLeaves, l)
+}
+
+//// Ratchet Tree
+type RatchetTree struct {
+	Nodes       []OptionalRatchetNode `tls:"head=2"`
+	CipherSuite CipherSuite
+	NumLeaves   leafCount
+}
+
+func newRatchetTree(cs CipherSuite) *RatchetTree {
+	return &RatchetTree{
+		Nodes:       []OptionalRatchetNode{},
+		CipherSuite: cs,
+	}
+}
+
+func (t *RatchetTree) Encap(from leafIndex, context, leafSecret []byte) (*DirectPath, []byte) {
+	dp := &DirectPath{}
+	leafNode := toNodeIndex(from)
+	n := t.newNode(leafSecret)
+	if t.Nodes[leafNode].Node != nil {
```

Comment:
You can avoid this branch by moving the `if n == nil` check into `Merge()`

Reply:
Actually, no, don't do that.  You should never Encap() from an unoccupied leaf, because then you don't have a credential to sign with!